### PR TITLE
feat(undici): add `ProxyAgent` for `undici`

### LINKED ISSUE #4457: `ProxyAgent` in `"undici"` doesn't work
### What version of Bun is running?

0.8.2+5288178c865cd9dcb500b00a65197bf2641d4e5c

### What platform is your computer?

Darwin 22.4.0 arm64 arm

### What steps can reproduce the bug?

hi, I got some problem with undici

I've already get undici installed by bun
```
╰>>> bun add undici
bun add v0.8.2 (5288178c)

 installed undici@5.23.0
```

and here is the sample code 
```ts
import { ProxyAgent } from "undici";

console.log(ProxyAgent.name);

```

but bun reports me the export class `ProxyAgent` not found in undici
```
╰>>> bun run index.ts 
SyntaxError: Import named 'ProxyAgent' not found in module 'undici'.
```

### What is the expected behavior?

ts-node can run this piece of code properly, print the name of the imported class `ProxyAgent`

```
╰>>> ts-node-esm index.ts 
ProxyAgent
```

### What do you see instead?

```
╰>>> bun run index.ts 
SyntaxError: Import named 'ProxyAgent' not found in module 'undici'.
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ThatOneBro**
### What does this PR do?

This PR adds `ProxyAgent` to the `undici` module, as well as converts `undici.js` -> `undici.ts`. 
There are likely a few bug fixes along with this feature but I still need to add more tests.

Should close #4457 and partially addresses #4474 with the caveat that I don't think TLS is supported yet 

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

TODO: 
- [ ] Write some tests for proxies

### CODE REVIEW

File: src/js/thirdparty/undici.ts [hunk 1]
```diff
@@ -102,36 +101,37 @@ class BodyReadable extends ReadableFromWeb {
 // *   upgrade?: boolean | string | null;
 // *   blocking?: boolean;
 
+type RequestOptions = {
+  dispatcher?: Dispatcher;
+  method: HttpMethod | string;
+  signal?: AbortSignal | e.EventEmitter | null;
+  maxRedirections?: number;
+  body?: string | Buffer | Uint8Array | s.Readable | null | FormData;
+  headers?: IncomingHttpHeaders | string[] | null;
```
Comment (Hebilicious) [inline src/js/thirdparty/undici.ts:110]:
```suggestion
  headers?: IncomingHttpHeaders | null;
```

File: src/js/thirdparty/undici.ts [hunk 2]
```diff
@@ -200,21 +202,39 @@ async function request(
...
 
+  let proxy: string | undefined;
+  if (dispatcher && dispatcher instanceof ProxyAgent) {
+    // get proxy uri
+    proxy = dispatcher[kBunUndiciProxyAgentUri];
+    // if proxyAgent.token is defined, put a proxy auth header with token in the headers
+    const token = dispatcher[kBunUndiciProxyAgentToken];
+    if (token) {
+      if (inputHeaders) {
+        if (Array.isArray(inputHeaders)) {
```
Comment (Hebilicious) [inline src/js/thirdparty/undici.ts:213]: How can headers be `string[]` ? 🤔

File: src/js/thirdparty/undici.ts [hunk 3]
```diff
@@ -265,6 +285,63 @@ function Undici() {
...
+    if (uri.startsWith("https://")) throw new TypeError("proxy over TLS not implemented");
+    if (!uri.startsWith("http://")) throw new Error("uri must be a valid URL string with the protocol `http://`");
+
+    let token;
+    if (options) {
+      if (typeof options !== "object") throw new TypeError("options must be an object");
+      // Auth token can either be passed via `token` or `auth` prop, though `auth` is now deprecated
+      const { token: _token, auth } = options;
+      token = _token ?? auth;
+      if (token && (typeof token !== "string" || !token.startsWith("Basic ")))
```
Comment (Hebilicious) [inline src/js/thirdparty/undici.ts:317]: looking at undici implementation https://github.com/nodejs/undici/blob/83cb522ae0157a19d149d72c7d03d46e34510d0a/lib/proxy-agent.js#L72
`auth` should always need `Basic`, maybe error before and say to use token ?

File: src/js/thirdparty/undici.ts [hunk 4]
```diff
@@ -265,6 +285,63 @@ function Undici() {
...
+  token?: string;
+  headers?: IncomingHttpHeaders;
+};
+
+class ProxyAgent extends Dispatcher {
+  #uri: string;
+  #token?: string;
+  // #headers?: Dict<string | string[]>;
+
+  constructor(uri: string, options: ProxyAgentOptions) {
```
Comment (Hebilicious) [inline src/js/thirdparty/undici.ts:304]: `ProxyAgent` should also accept an object as a single argument

File: src/js/thirdparty/undici.ts [hunk 5]
```diff
@@ -265,6 +285,63 @@ function Undici() {
...
+
+  get [kBunUndiciProxyAgentUri]() {
+    return this.#uri;
+  }
+
+  get [kBunUndiciProxyAgentToken]() {
+    return this.#token;
+  }
+
+  // get [kBunUndiciProxyAgentHeaders]() {
```
Comment (Hebilicious) [inline src/js/thirdparty/undici.ts:334]: If think headers should be merged with the incoming headers, but if we can't maybe throw a notImplemented error/warning ?

**Review by Hebilicious — COMMENTED** [review]
Thank you for implementing this ❤️ 
I left a few comments here and there.
I'm also thinking it could be nice to emit a warning/have some TSdoc that inform the user about `fetch(url, { proxy })` when using `fetch(url, { dispatcher })` ?

File: src/js/thirdparty/undici.ts (hunk 2 above)
Comment (ThatOneBro) [reply src/js/thirdparty/undici.ts:213]: Good question, this is definitely wrong. I think it's actually supposed to be a `string[][]` or `[string, string][]` more precisely. This when headers are provided as an array of tuples of `["header-name", "header-value"]`. I'm going to fix this type and some other things later.

File: src/js/thirdparty/undici.ts (hunk 3 above)
Comment (ThatOneBro) [reply src/js/thirdparty/undici.ts:317]: 100% I missed that, good catch

**Comment by Zamion101** [comment]
Any update on this PR? We really need undici Proxies for sending requests.

**Comment by Jarred-Sumner** [comment]
Closing due to bitrot

**Pull request closed without merging**